File: leet_BTREE_606_btree_to_string___preorder.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    string tree2str(TreeNode* root)
    {
        vector<string>  res;
        string          r;

        preorder(root, res);
        for (string s: res)
            r += s;
        return r.substr(1, r.length() - 2);
    }

    void    preorder(TreeNode *node, vector<string>& res)
    {
        if (!node)
            return ;
        res.push_back("(");
        res.push_back(to_string(node->val));
        if (! node->left && node->right)
        {
            res.push_back("()");
        }
        preorder(node->left, res);
        preorder(node->right, res);
        res.push_back(")");
    }
};

File: leet_BTREE_606_btree_to_string___preorder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def tree2str(self, root: Optional[TreeNode]) -> str:
        a = []
        def preorder(node):
            if not node:
                return
            a.append('(')
            a.append(str(node.val))
            if not node.left and node.right:
                a.append('()')
            preorder(node.left)
            preorder(node.right)
            a.append(')')
        preorder(root)
        return ''.join(a[1:-1])
